src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+/// for finished log segments, converts them to `Batch`s (basically consolidates the
+/// updates for a range of times) and adds them to the relation's `Trace`.
+///
+/// When a `Trace` contains too many `Batches`, the Trace physically
+/// combines all of them into a single large batch with updates
+/// forwarded to the compaction frontier.
+
+// TODO: Lets add some jitter to compaction so we aren't compacting every single
+// relation at the same time maybe?
+static COMPACTER_INTERVAL: Duration = Duration::from_secs(300);
+
+/// Instructions that the Coordinator sends the Compacter.
+#[derive(Debug)]
+pub enum CompacterMessage {
+    Add(GlobalId),
+    Drop(GlobalId),
+    Resume(GlobalId, Trace),
+    AllowCompaction(GlobalId, Timestamp),
+}
+
+/// A Batch contains all of the updates that originated within some time range [lower, upper)
+/// but the data live the file at `path`.
+///
+/// The data stored in each batch are triples of (Row, Timestamp, Diff) such that there is
+/// exactly one copy of each (Row, Timestamp) in each batch. Batches also have a header
+/// and a footer indicating the upper and lower bound timestamps.
+/// TODO: Batches are meant to mimic differential / cdcv2 batches but do not do so
+/// currently. Let's fix that. Specifically, introduce a `since` field, and counts
+/// for the number of updates at each time.
+#[derive(Debug)]
+struct Batch {
+    upper: Timestamp,
+    lower: Timestamp,
+    path: PathBuf,
+}
+
+impl Batch {
+    /// Create a batch from a finished log segment file.
+    fn create(log_segment_path: &Path, trace_path: &Path) -> Result<Self, anyhow::Error> {
+        let messages = read_segment(log_segment_path)?;
+        Batch::create_from_messages(messages, trace_path, None)
+    }
+
+    /// Read in and consolidate a list of messages, and write them to a new batch file
+    /// in `trace_path`.
+    ///
+    /// Will also compact updates up to `compaction_frontier` if provided..
+    fn create_from_messages(
+        messages: Vec<Message>,
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut upper: Option<Timestamp> = None;
+        let mut lower: Option<Timestamp> = None;
+        let mut time_data = BTreeMap::new();
+        for message in messages.iter() {
+            match message {
+                Message::Progress(time) => match (lower, upper) {
+                    (None, None) => {
+                        lower = Some(*time);
+                    }
+                    (Some(l), None) => {
+                        assert!(*time >= l);
+                        upper = Some(*time);
+                    }
+                    (Some(_), Some(u)) => {
+                        assert!(*time >= u);
+                        upper = Some(*time);
+                    }
+                    (None, Some(_)) => unreachable!(),
+                },
```

Comment:
What's going on here? I could imagine some "figure out and update some bounds" but best to spell it out, as it isn't otherwise evident.

Reply:
For example, idk if seeing the first `Message::Progress` is actually evidence that it is the lower bound. Could data have arrived before that message, at an earlier time?

Reply:
added some more documentation + an assert whenever we see data that a lowerbound is present.